README.md
# pyhetdex

`pyhetdex` is a python library designed to collect code can could be used by
different project like
[`VIRUS Data Analysis Tool`](https://luna.mpe.mpg.de/wikihetdex/index.php/VIRUS_Data_Analysis_Tool),
[`VIRUS Health Check`](https://luna.mpe.mpg.de/wikihetdex/index.php/VIRUS_Health_Check),
[`simhetdex`](https://luna.mpe.mpg.de/wikihetdex/index.php/VIRUS_Data_Simulation_Framework)
or
[`Cure-WISE`](https://luna.mpe.mpg.de/wikihetdex/index.php/Overview_of_Cure-WISE).

## Official documentation

The official documentation can be found
[online](http://www.mpe.mpg.de/~montefra/documentation/pyhetdex).

We use the [Redmine issue
tracker](https://luna.mpe.mpg.de/redmine/projects/pyhetdex) to register bugs,
feature requests and other issues.

## Installation

The recommended way to install pyhetdex is using
[pip](https://pip.pypa.io/en/latest/):

    pip install --extra-index-url https://gate.mpe.mpg.de/pypi/simple/ pyhetdex

It's possible to set the extra index URL permanently by adding the following
lines to the ``$HOME/.pip/pip.conf`` file:

    [global]
    extra-index-url = https://gate.mpe.mpg.de/pypi/simple

or exporting the environment variable:

    export PIP_EXTRA_INDEX_URL=https://gate.mpe.mpg.de/pypi/simple

The list of released versions can be seen [on the MPE pypi
server](https://gate.mpe.mpg.de/pypi/simple/pyhetdex/). A specific version can
be installed using
[specifiers](https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers>),
e.g. issuing ``pip install pyhetdex==0.5``.

``pip`` will take care of installing all the pyhetdex dependences

We suggest you install pyhetdex into a
[virtualenv](https://virtualenv.pypa.io/en/stable/), in an
[anaconda](https://www.continuum.io/downloads)/[conda](http://conda.pydata.org/docs/index.html)
or in similar environments.

Of course it is also possible to install pyhetdex without any of the above with::

    pip install --user --extra-index-url https://gate.mpe.mpg.de/pypi/simple/ pyhetdex

This way the pyhetdex executables are installed in ``$HOME/.local/bin``, so make
sure to add this to the environment variable ``PATH`` to be able to easily use
them on the command line. The use of ``sudo`` when installing with pip is
[discouraged](http://stackoverflow.com/questions/21055859/what-are-the-risks-of-running-sudo-pip)
and potentially harmful.

## Dependencies

Mandatory

    six
    numpy
    matplotlib
    scipy
    astropy>=1
    Pillow

Optional:

    test: pytest-cov
          pytest-xdist
          pytest-catchlog
          pytest
          tox
    documentation: sphinx
                   numpydoc
                   alabaster

## Development

If you want to help developing pyhetdex, please follow the information about
[installation](http://www.mpe.mpg.de/~montefra/documentation/pyhetdex/latest/install.html#development)
and
[contribution](http://www.mpe.mpg.de/~montefra/documentation/pyhetdex/latest/contributions.html)
and get in touch with the other developers.

## Notes

Please use [markdown](http://daringfireball.net/projects/markdown/syntax) syntax
so it can be easily converted to html or wiki format if needed **and** remains
very readable



Changelog
2021-10-07 Daniel Farrow <dfarrow@mpe.mpg.de>

    * pyhetdex/tools/read_catalogues.py: Change xoff, yoff defaults again

2019-01-15 Daniel Farrow <dfarrow@mpe.mpg.de>

    * pyhetdex/tools/read_catalogues.py: Change xoof, yoff defaults

2018-07-30 Daniel Farrow <dfarrow@mpe.mpg.de>

       * pyhetdex/coordinates/astrometry.py: Completely removed rho
         stage rotation option
       * tests/test_coordinates/test_astrometry.py: Removed tests
         of rho stage rotation


2018-07-27  Jan Snigula  <snigula@mpe.mpg.de>

	* tests/data/20180522T023753.4_053*_zro.fits: Added files needed
	for pytest

2018-06-04 Daniel Farrow <dfarrow@mpe.mpg.de>

        * pyhetdex/coordinates/astrometry.py: Changed default rho stage
	rotation to zero
        * tests/test_coordinates/test_astrometry.py: Updated the tests for
	the above change

2018-06-04  Jan Snigula  <snigula@mpe.mpg.de>

	* setup.py: Bumped version to 0.14.2-post

2018-06-04  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/overview_plot.py: Added close() function to clean
	up the figure object and clear the figure.
	* setup.py: Bumped version to 0.14.2

2018-05-28  Jan Snigula  <snigula@mpe.mpg.de>

	* setup.py: Bumped version to 0.14.1-post

2018-05-28  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/overview_plot.py: Do explicit dels and a clf to
	clear up used memory by the plotting, and fix a bug that caused
	only one quadrant per IFU to be plotted
	* setup.py: Bumped version to 0.14.1

2018-05-22  Jan Snigula  <snigula@mpe.mpg.de>

	* setup.py: Bumped version to 0.14.0-post

2018-05-22  Jan Snigula  <snigula@mpe.mpg.de>

	* ReleaseNotes.rst: Updated for release
	* setup.py: Bumped version to 0.14.0

2018-05-22  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/check_repeat.py: Added documentation

2018-05-18  Jan Snigula  <snigula@mpe.mpg.de>

	* ReleaseNotes.rst: Updated release notes

2018-05-16 Francesco Montesano <montefra@mpe.mpg.de>

    * readthedocs.yml: added
    * requirements.txt: removed

2018-05-15 Francesco Montesano <montefra@mpe.mpg.de>

    * doc/source/index.rst: add the readthedocs badge and remove coverage link

2018-05-15 Francesco Montesano <montefra@mpe.mpg.de>

    * ReleaseNotes.rst: converted from markdown
    * doc/source/release_notes.rst: added
    * doc/source/index.rst: add the above

2018-05-14 Jan Snigula <snigula@mpe.mpg.de>

    * Changelog: Dummy change to test documentation hook

2018-05-14 Francesco Montesano <montefra@mpe.mpg.de>

    * requirements.txt: renamed and add all dependences

2018-05-14 Francesco Montesano <montefra@mpe.mpg.de>

    * requirements_docs.txt: added

2018-05-08 Francesco Montesano <montefra@mpe.mpg.de>

    * pyhetdex/**.py: add the full copyright notice

2018-05-07 Francesco Montesano <montefra@mpe.mpg.de>

    * LICENSE: rename and add GPL3
    * doc/source/licence.rst: update
    * AUTHORS: fix typo
    * doc/source/conf.py: update copyright
    * pyhetdex/**.py: add copyright notice

2018-03-13  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/check_repeat.py: Added more output
	* pyhetdex/het/overview_plot.py: Added draw() statement before
	saving the image, fixing Issue #2419.

2018-03-12  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/check_repeat.py: Change function parameters, to
	make it easier to call it from hetwise ingestion.

2018-03-12  Jan Snigula  <snigula@mpe.mpg.de>

	* pyhetdex/het/check_repeat.py: Added tool to check data for
	repeat data. This checks the md5 sum of the fits data section.

	* setup.py: Added check_repeat entry point

2018-02-08 Francesco Montesano <montefra@mpe.mpg.de>

    * setup.py: bump version to 0.13.0-post

2018-02-08 Francesco Montesano <montefra@mpe.mpg.de>

    * setup.py: prepare of v0.13.0 release
    * ReleaseNotes.md: same

2018-02-08 Francesco Montesano <montefra@mpe.mpg.de>

    * : merge ^/branches/peewee3/ into ^/trunk

2018-02-08 Francesco Montesano <montefra@mpe.mpg.de>

    * pyhetdex/tools/db_helpers.py: peewee 3 changed the method to get the
	  connection (resolves issue #2344)
    * tox.ini: test with peewee 2
    * ReleaseNotes.md: udpate

2018-01-19 Francesco Montesano <montefra@mpe.mpg.de>

	* ReleaseNotes.md: update with forgotten changes

2018-01-19 Francesco Montesano <montefra@mpe.mpg.de>

	* : merge ^/branches/skip_comments into ^/trunk
	* ReleaseNotes.md: update

2018-01-18 Francesco Montesano <montefra@mpe.mpg.de>

    * tests/test_tools/test_file_tools.py: make sure that files will all
      comments are properly handled. Resolves #2305

2018-01-18 Francesco Montesano <montefra@mpe.mpg.de>

    * pyhetdex/tools/files/file_tools.py: rewrite skip_comments: implements issue
	  #2305
    * tests/data/IFUcen_HETDEX_dos.txt: add file with dos line ends for testing
    * tests/test_tools/test_file_tools.py: test the changes
    * tests/test_het/test_ifu_centers.py: test if the IFUcen file parser is
	  more robust

2017-12-13 Francesco Montesano <montefra@mpe.mpg.de>

    * tox.ini: use pytest>3.3 and remove pytest-catchlog. Resolves #2256
    * setup.py: same
    * doc/source/install.rst: update accordingly
    * doc/Makefile: open the browser
    * doc/source/contributions.rst: update accordingly

2017-12-13 Francesco Montesano <montefra@mpe.mpg.de>

	* : merge ^/branches/remove_deprecated2 into ^/trunk

2017-12-13 Francesco Montesano <montefra@mpe.mpg.de>

	* : merge ^/trunk into ^/branches/remove_deprecated2

2017-09-19 Francesco Montesano <montefra@mpe.mpg.de>

    * pyhetdex/tools/configuration.py: remove deprecated classes. Resolves issue
	  #1981
    * tests/test_tools/test_configuration.py: update the tests

2017-09-19 Francesco Montesano <montefra@mpe.mpg.de>

    * pyhetdex/het/fplane.py: remove deprecated ``difus`` property. Resolves
	  issue #2120
    * doc/source/het/fplane.rst: update the docs
    * pyhetdex/ltl/marray.py: forgot to remove one import

2017-09-19 Francesco Montesano <montefra@mpe.mpg.de>


pyhetdex/ltl/MArray.py
from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

__version__ = '$Id$'

import numpy as np
from pyhetdex.tools import io_helpers

_typedict = {'float' : 'single', 'double' : 'double', 'int' : 'intc', 'char' : 'byte'}
_formatdict = {'float' : ' %.16e', 'double' : ' %.16e', 'int' : ' %d', 'char' : ' %c'}

class FVector:
    ''' Python class allowing reading / writing of ltl/FVector object.'''
    
    data=None
    ftype=''
    format=''
    size=0
    
    def __init__(self,dtype='float'):
        if not _typedict.has_key(dtype):
            raise Exception( 'Data format %s is not supported.' % dtype ) 
        self.ftype = _typedict[dtype]
        self.format = _formatdict[dtype]
            
    def read(self, ios):
        
        # First read the header, format is:
        # FVector< T,36,0 >
        
        dtype = io_helpers.read_to_char( ios, '<' ).strip()
        if ( dtype != 'FVector'):
            raise Exception( 'Expected FVector, got %s'  % dtype)
        
        io_helpers.eat_to_char( ios, ',') # Forward through  < T,

        self.size = int( io_helpers.read_to_char( ios, ',') ) # Read size to next ,
        
        stride = int( io_helpers.read_to_char( ios, '>') )
        if ( stride != 0 ):
            raise Exception( 'Only stride 0 is supported' )
        
        # Read actual data in
        # Data is in format  [ 6.4277265720642998e+00 4.6045159399542094e+01 1.7884131502943873e+02 ]
        # Could span several lines
        io_helpers.eat_to_blockstart( ios ) # Go to start of data
            
        self.data = np.fromstring(io_helpers.read_to_char( ios, ']'), dtype=self.ftype, sep=' ')
                
        if ( self.data.size != self.size ):
            raise Exception( 'Expected to read %d elements, got %d' % (self.size, self.data.size))
        
        # Clean up the trainling newline
        io_helpers.eat_to_char(ios, '\n')
    
    def write(self, ios):
        
        ios.write( "FVector< T,%d,0 >\n " % self.size )
        ios.write( "[" )
        for i in range( 0, self.size ):
            ios.write( self.format % self.data[i] )
            if not ((i+1)%9):
                ios.write( '\n  ' )

        ios.write( " ]\n")
        
class MArray:
    ''' Python class allowing reading / writing of ltl/MArray object.'''
    
    data=None
    ftype=''
    format=''
    ndims=0
    size=0
    
    def __init__(self,dtype='float'):
        if not _typedict.has_key(dtype):
            raise Exception( 'Data format %s is not supported.' % dtype ) 
        self.ftype = _typedict[dtype]
        self.format = _formatdict[dtype]
            
    def read(self, ios):
        
        # First read the header, format is:
        # MArray<T,2> ( 14 x 246 ) : (1,14) (1,246)
        
        dtype = io_helpers.read_to_char( ios, '<' ).strip()
        if ( dtype != 'MArray'):
            raise Exception( 'Expected MArray, got %s'  % dtype)
        
        io_helpers.eat_to_char( ios, ',') # Forward through  < T,

        self.ndims = int( io_helpers.read_to_char( ios, '>') ) # Read dimensions
        
        io_helpers.eat_to_char( ios, '(') # Forward to sizes
        
        self.size=[]

        d=1        
        while ( d < self.ndims ):
            self.size.append( int( io_helpers.read_to_char( ios, 'x' )))
            d+=1

        self.size.append( int( io_helpers.read_to_char( ios, ')' ))) 

        #ios.readline() # Read rest of descriptor line
        
        # Read actual data in
        # Data is in format  [ 6.4277265720642998e+00 4.6045159399542094e+01 1.7884131502943873e+02 ]
        # Could span several lines

        # Allocate data array
        self.size.reverse()
        data = np.zeros( self.size, self.ftype)
        self.size.reverse()
        
        # Read actual data in
        io_helpers.eat_to_blockstart( ios )
        
        # Now we are at the beginning of the first data block
        
        i=0
        while ( i < data.size ):
            val = io_helpers.read_to_char( ios, ']' )
            data.flat[i:i+self.size[0]] = np.fromstring( val, dtype=self.ftype, sep=' ' )
            i = i+self.size[0] 
            if( i < data.size ):
                io_helpers.eat_to_blockstart( ios )
        
        self.data = data.transpose()
        
        # Clean up the trailing newline
        io_helpers.eat_to_char( ios,'\n')    

    def write(self, ios):

        # Write header line
        dimstr='( %d' % self.size[0]
        sizestr='(1,%d)' % self.size[0]
        
        d=1        
        while ( d < self.ndims ):
            dimstr = dimstr + (" x %d" % self.size[d] )
            sizestr = sizestr + (" (1,%d)" % self.size[d])
            d+=1
        dimstr = dimstr + " )"

        ios.write( "MArray<T,%d> %s : %s\n" % (self.ndims,dimstr,sizestr) )
        
        # Write the data
        
        self.__recursive_write( ios, self.data, 0, ' ')
        ios.write(']\n')

    def __recursive_write(self, ios, data, i, pad):
        ios.write( '[' )
        if (data.ndim > 1):
            while ( i < data.shape[-1]):
                self.__recursive_write( ios, data[...,i], 0, pad+' ')
                i = i+1
                ios.write( ']' )
                if( i < data.shape[-1]):
                    ios.write( '\n'+pad)
        else:
            for j in range( 0, data.size ):
                ios.write( self.format % data[j] )
                if not ((j+1)%9):
                    ios.write( '\n'+pad )
            ios.write( ' ' )

def interpCheby2D_7(x, y, p):

    if isinstance(x, (tuple, list)):
        x = np.asarray(x)
    if isinstance(y, (tuple, list)):
        y = np.asarray(y)
    
    T2x = 2.  * x**2 -  1.
    T3x = 4.  * x**3 -  3.   * x
    T4x = 8.  * x**4 -  8.   * x**2 + 1.
    T5x = 16. * x**5 -  20.  * x**3 + 5. * x
    T6x = 32. * x**6 -  48.  * x**4 + 18. * x**2 - 1.
    T7x = 64. * x**7 -  112. * x**5 + 56. * x**3 - 7. * x
    T2y = 2.  * y**2 -  1.;    
    T3y = 4.  * y**3 -  3.   * y
    T4y = 8.  * y**4 -  8.   * y**2 + 1.
    T5y = 16. * y**5 -  20.  * y**3 + 5. * y
    T6y = 32. * y**6 -  48.  * y**4 + 18. * y**2 - 1
    T7y = 64. * y**7 -  112. * y**5 + 56. * y**3 - 7 * y

    return  p[0]*T7x    + p[1]*T6x    + p[2]*T5x      + p[3]*T4x      + p[4]*T3x      + p[5]*T2x      + p[6]*x + \
            p[7]*T7y    + p[8]*T6y    + p[9]*T5y      + p[10]*T4y     + p[11]*T3y     + p[12]*T2y     + p[13]*y + \
            p[14]*T6x*y + p[15]*x*T6y + p[16]*T5x*T2y + p[17]*T2x*T5y + p[18]*T4x*T3y + p[19]*T3x*T4y + \
            p[20]*T5x*y + p[21]*x*T5y + p[22]*T4x*T2y + p[23]*T2x*T4y + p[24]*T3x*T3y + \
            p[25]*T4x*y + p[26]*x*T4y + p[27]*T3x*T2y + p[28]*T2x*T3y + \
            p[29]*T3x*y + p[30]*x*T3y + p[31]*T2x*T2y + \
            p[32]*T2x*y + p[33]*x*T2y + p[34]*x*y     + p[35]

